src/solver/opamCudf.ml

```diff
@@ -677,29 +677,35 @@ let arrow_concat sl =
   in
   String.concat (OpamConsole.colorise `yellow arrow) sl
 
-let formula_of_vpkgl cudfnv2opam all_packages vpkgl =
-  let atoms =
-    List.map (fun vp ->
-        try vpkg2atom cudfnv2opam vp
-        with Not_found ->
-          OpamPackage.Name.of_string (Dose_common.CudfAdd.decode (fst vp)), None)
+let formula_of_vpkgl cudfnv2opam vpkgl =
+  let formula_map =
+    List.fold_left (fun acc vp ->
+        let name, version_constraint =
+          try vpkg2atom cudfnv2opam vp
+          with Not_found ->
+            (OpamPackage.Name.of_string (Dose_common.CudfAdd.decode (fst vp)), None)
+        in
+        let version_formula = match version_constraint with
+          | Some version_constraint -> OpamFormula.Atom version_constraint
+          | None -> OpamFormula.Empty
+        in
+        let update prev =
+          OpamFormula.Or (prev, version_formula)
+        in
+        OpamPackage.Name.Map.update name update OpamFormula.Empty acc)
+      OpamPackage.Name.Map.empty
       vpkgl
   in
-  let names = OpamStd.List.sort_nodup compare (List.map fst atoms) in
-  let by_name =
-    List.map (fun name ->
-        let formula =
-          OpamFormula.ors (List.map (function
-              | n, Some atom when n = name -> Atom atom
-              | _ -> Empty)
-              atoms)
-        in
-        let all_versions = OpamPackage.versions_of_name all_packages name in
-        let formula = OpamFormula.simplify_version_set all_versions formula in
```

Comment:
This was the only use of that function. Do we want to also remove it from `OpamFormula`? Given this function obviously has a serious bug, should we also debug whether it comes from this one or `OpamFormula.gen_formula` as we might want to remove or fix this one too?

Reply:
I chose this route because trying to debug this code sounds like a time sink

Reply:
> obviously has a serious bug

can you be more explicit ?

I believe the issue may come not from the function itself, but from the fact that it generates a formula that results in the same partition of the set of packages *that it was supplied*. And that already excludes unavailable packages, which is where the results may become misleading once put in the final message.
This is all IIRC :grimacing:, it's been a while.

The fundamental issue, again IIRC, is that we are trying to extract the reasons from a graph of dependencies between individual package versions, at a point where the initial formulas have already been evaluated ; when walking the dependency chains we are performing set operations, etc. At this stage, recovering the initial formula is not really possible (and after set union, maybe not even desirable).

Of course here the focus case seems to be on singletons, where the results seem very dumb: we may want to special-case this case of figure — I believe the formula generation function would favor inequalities when only the upper or lower bound matches, we could prefer equalitites (both being correct mathematically, either being potentially wrong w.r.t. the original formula, which we can't retrieve at this point).

I thought this was no longer the case, but at some point, when encoding the Cudf constraint, we would transpose a never-matching constraint with `< min-version` for lack of a better thing. This transpiring into the user messages is pretty bad.

> I chose this route because trying to debug this code sounds like a time sink

Yeah, I agree :sweat: